src/preprocess/mathjax.rs

```diff
@@ -0,0 +1,261 @@
+            (\\\\\)) # followed by a closing bracket `\\)`
+
+            |        # or
+
+                     # Legacy block mathematics
+            (\\\\\[) # An escaped opening bracket `\\[`
+            ([^$]+)  # followed by one ore more other things TODO provide correct regexp.
+            (\\\\\]) # followed by a closing bracket `\\]`
+        ").unwrap();
+    }
+    MathematicsIterator(REGEXP.captures_iter(content))
+}
+
+struct MathematicsIterator<'a>(CaptureMatches<'a, 'a>);
+
+impl<'a> Iterator for MathematicsIterator<'a> {
+    type Item = Mathematics<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        for capture in &mut self.0 {
+            if let mathematics @ Some(_) = Mathematics::from_capture(capture) {
+                return mathematics;
+            }
+        }
+        None
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct Mathematics<'a> {
+    start_index: usize,
+    end_index: usize,
+    kind: Kind,
+    text: &'a str,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum Kind {
+    Inline,
+    Block,
+    LegacyInline,
+    LegacyBlock,
+}
+
+impl<'a> Mathematics<'a> {
+    fn from_capture(captures: Captures<'a>) -> Option<Self> {
+        let kind =
+            captures.get(1).or(captures.get(4)).or(captures.get(7)).or(captures.get(10))
+            .map(|delimiter|
+                 match delimiter.as_str() {
+                     "$$"    => Kind::Block,
+                     "$"     => Kind::Inline,
+                     "\\\\[" => Kind::LegacyBlock,
+                     _       => Kind::LegacyInline,
+                 })
+            .expect("captured mathematics should have opening delimiter at the provided indices");
+
+        captures.get(0).map(|m| Mathematics {
+            start_index: m.start(),
+            end_index: m.end(),
+            kind: kind,
+            text: kind.text(m.as_str()),
+        })
+    }
+
+    fn replacement(&self) -> String {
+        let mut replacement = String::new();
+        match self.kind {
+            Kind::Block  | Kind::LegacyBlock  => {
+                replacement.push_str("<div class=\"math\">$$");
+                replacement.push_str(self.text);
+                replacement.push_str("$$</div>");
+            },
+            Kind::Inline | Kind::LegacyInline => {
+                replacement.push_str("<span class=\"inline math\">$");
+                replacement.push_str(self.text);
+                replacement.push_str("$</span>");
+            },
+        }
+        replacement
+    }
+}
+
+impl Kind {
+    fn text<'a>(&self, delimited_text: &'a str) -> &'a str {
+        let end = delimited_text.len();
+        match *self {
+            Kind::Block        => &delimited_text[2..end-2],
+            Kind::Inline       => &delimited_text[1..end-1],
+            Kind::LegacyBlock  => &delimited_text[3..end-3],
+            Kind::LegacyInline => &delimited_text[3..end-3],
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
As you've mentioned, we probably want to flesh these test out a bit. Off the top of my head we probably want to check

- Strings with just one `$`
- What happens when you have a newline between delimiters (e.g. `"$foo\nbar$`)
- Is there a limit to the length of an expression? (e.g. you have a `$` at the start of your chapter and another towards the end of the page)
- What happens if I overlap delimiters? e.g. `$a^{2} + b^{2} = c^{2}\\]` and so on.

Reply:
I was thinking of these kinds of tests. I will add them in.